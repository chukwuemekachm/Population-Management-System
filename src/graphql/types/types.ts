import { Request } from 'express-serve-static-core';

import { Prisma } from '../prisma/generated/prisma-client';

// Data Types
export type DateTime = string;
export type ID = string;

// Resolver Types
export type AuthResponseResolver = (parent: AuthResponse) => AuthResponse;
export type UserResolver = (parent: User) => User;
export type LocationResolver = (parent: Location) => Location;

// Common Types
export interface Context {
  request: Request | {};
  prisma: Prisma;
}

export interface AuthResponse {
  token: string;
  user: Partial<User>;
}

// Model Types
export interface User {
  id: ID;
  firstName: string;
  lastName: string;
  email: string;
  locations: Location[];
  createdAt: DateTime;
  updatedAt: DateTime;
}

export interface Location {
  id: ID;
  name: string;
  maleResidents: number;
  femaleResidents: number;
  totalResidents: number;
  parentLocationId: ID | null;
  subLocations: Location[];
  creator: User;
  createdAt: DateTime;
  updatedAt: DateTime;
}
