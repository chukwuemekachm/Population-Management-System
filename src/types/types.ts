import { Request } from 'express-serve-static-core';

import { Prisma } from '../prisma/generated/prisma-client';

export type DateTime = string;
export type ID = string;

export interface Context {
  request: Request;
  prisma: Prisma;
}

export interface User {
  id: ID;
  firstName: string;
  lastName: string;
  email: string;
  locations: Location[];
  createdAt: DateTime;
  updatedAt: DateTime;
}

export interface Location {
  id: ID;
  name: string;
  maleResidents: number;
  femaleResidents: number;
  totalResidents: number;
  parentLocationId: ID;
  subLocations: Location[];
  creator: User;
  createdAt: DateTime;
  updatedAt: DateTime;
}
