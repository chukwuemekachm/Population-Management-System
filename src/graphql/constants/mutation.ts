import { gql } from 'apollo-server-express';

export const SIGNUP_MUTATION = gql`
  mutation signup($user: SignupInput) {
    signup(user: $user) {
      user {
        email
        firstName
        lastName
      }
    }
  }
`;

export const LOGIN_MUTATION = gql`
  mutation login($user: LoginInput) {
    login(user: $user) {
      user {
        firstName
        lastName
      }
    }
  }
`;
