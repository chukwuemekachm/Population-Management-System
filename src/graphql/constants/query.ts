import { gql } from 'apollo-server-express';

export const GET_LOCATIONS_QUERY = gql`
  query getLocations($filter: LocationFilterInput) {
    getLocations(filter: $filter) {
      locationName
    }
  }
`;

export const GET_LOCATION_QUERY = gql`
  query getLocation($locationId: String!) {
    getLocation(locationId: $locationId) {
      locationName
    }
  }
`;
