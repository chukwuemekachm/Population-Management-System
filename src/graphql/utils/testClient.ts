import { ApolloServer } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';

import typeDefs from '../schema';
import resolvers from '../resolvers';
import prismaTestMockClient from '../__mocks__/prisma';
import { User } from '../types/types';

export const constructGraphqlTestClient = (user?: Partial<User>) => {
  const testServer = new ApolloServer({
    typeDefs,
    resolvers,
    context: ({ req }) => ({
      request: req,
      prisma: prismaTestMockClient,
      user,
    }),
  });

  return createTestClient(testServer);
};
