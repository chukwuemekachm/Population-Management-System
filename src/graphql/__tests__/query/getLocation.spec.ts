import prismaTestMockClient from '../../__mocks__/prisma';
import { GET_LOCATION_QUERY } from '../../constants';
import { userInputs } from '../../fixtures/userFixtures';
import { constructGraphqlTestClient } from '../../utils/testClient';
import { User } from '../../types/types';
import { locationInputs } from '../../fixtures/locationFixtures';

let user: Partial<User>;
let locationId: string;

describe('get single location query', () => {
  beforeAll(async () => {
    const { mutate } = constructGraphqlTestClient();
    await prismaTestMockClient.deleteManyLocations();
    await prismaTestMockClient.deleteManyUsers();
    user = await prismaTestMockClient.createUser(userInputs[1]);
    ({ id: locationId } = await prismaTestMockClient.createLocation({
      locationName: locationInputs[0].locationName,
      femalePopulation: locationInputs[0].femalePopulation,
      malePopulation: locationInputs[0].malePopulation,
      creator: {
        connect: {
          id: user.id,
        },
      },
    }));
  });
  afterAll(async () => {
    await prismaTestMockClient.deleteManyLocations();
    await prismaTestMockClient.deleteManyUsers();
  });

  test('should return a single location matching the id on the platform', async () => {
    const { query } = constructGraphqlTestClient(user);
    const payload = {
      variables: {
        locationId,
      },
    };
    const { errors, data } = await query({
      mutation: GET_LOCATION_QUERY,
      ...payload,
    });
    expect(errors).toBeUndefined();
    expect(data).toBeDefined();
    expect(data).toMatchSnapshot();
  });

  test('should return a null when platform does not exist', async () => {
    const { query } = constructGraphqlTestClient(user);
    const payload = {
      variables: {
        locationId: 'jnsu9zi0o0jh89io0o0zmi8n89',
      },
    };
    const { errors } = await query({
      mutation: GET_LOCATION_QUERY,
      ...payload,
    });
    expect(errors).toBeDefined();
    expect(errors).toMatchSnapshot();
  });
});
