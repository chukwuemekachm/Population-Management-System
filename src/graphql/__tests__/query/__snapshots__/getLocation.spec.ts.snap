// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get single location query should return a null when platform does not exist 1`] = `
Array [
  [GraphQLError: Location with id: jnsu9zi0o0jh89io0o0zmi8n89 does not exist],
]
`;

exports[`get single location query should return a single location matching the id on the platform 1`] = `
Object {
  "getLocation": Object {
    "locationName": "",
  },
}
`;
